pypdf/_cmap.py

```diff
@@ -423,57 +433,93 @@ def compute_space_width(
             st = w[0] if isinstance(w[0], int) else w[0].get_object()
             second = w[1].get_object()
             if isinstance(second, int):
-                for x in range(st, second):
-                    w1[x] = w[2]
+                # C_first C_last same_W
+                en = second
+                for c_code in range(st, en + 1):
+                    try:
+                        conversion_char = map_dict[chr(c_code)]
+                        font_width_map[ord(conversion_char)] = w[2]
+                    except Exception:
+                        pass
                 w = w[3:]
             elif isinstance(second, list):
-                for y in second:
-                    w1[st] = y
-                    st += 1
+                # Starting_C [W1 W2 ... Wn]
+                c_code = st
+                for width in second:
+                    try:
+                        conversion_char = map_dict[chr(c_code)]
+                        font_width_map[ord(conversion_char)] = width
+                    except Exception:
+                        pass
+                    c_code += 1
                 w = w[2:]
             else:
                 logger_warning(
                     "unknown widths : \n" + (ft1["/W"]).__repr__(),
                     __name__,
                 )
                 break
-        try:
-            sp_width = w1[space_code]
-        except Exception:
-            sp_width = (
-                w1[-1] / 2.0
-            )  # if using default we consider space will be only half size
     elif "/Widths" in ft:
-        w = list(ft["/Widths"])  # type: ignore
-        try:
-            st = cast(int, ft["/FirstChar"])
-            en: int = cast(int, ft["/LastChar"])
-            if st > space_code or en < space_code:
-                raise Exception("Not in range")
-            if w[space_code - st].get_object() == 0:
-                raise Exception("null width")
-            sp_width = w[space_code - st].get_object()
-        except Exception:
-            if "/FontDescriptor" in ft and "/MissingWidth" in cast(
-                DictionaryObject, ft["/FontDescriptor"]
-            ):
-                sp_width = ft["/FontDescriptor"]["/MissingWidth"].get_object()  # type: ignore
-            else:
-                # will consider width of char as avg(width)/2
-                m = 0
-                cpt = 0
-                for xx in w:
-                    xx = xx.get_object()
-                    if xx > 0:
-                        m += xx
-                        cpt += 1
-                sp_width = m / max(1, cpt) / 2
-
-    if is_null_or_none(sp_width):
-        sp_width = 0.0
+        w = list(ft["/Widths"]) if isinstance(ft["/Widths"], ArrayObject) else []
```

Comment:
What happens if `ft["/Widths"]` is a `IndirectObject` pointing to an `ArrayObject`? AFAIK we should use `ft["/Widths"].get_object()` here.

Reply:
I checked the source code.
Is it correct that `IndirectObject` can take all values? It is written out once with `write_to_stream`, and if the content is `str` or `bytes`, it is converted to int, and if that fails, an error occurs.
From looking at the source code, it seems like there is no other way to handle it. I'm probably wrong, but I don't know the correct way to handle it.

Reply:
`IndirectObject` is basically just a pointer to any other PDF object, identified by its numerical ID.

Which exception would we handle here? In my opinion,

```python3
        w = ft["/Widths"].get_object()
```

should be sufficient. Why?

* This branch is only executed when `if "/Widths" in ft` holds true. Thus, we cannot really get a key error here.
* `ft["/Widths"]` returns a `PdfObject`. Each `PdfObject` supports `get_object(self)`, returning `self` by default. `IndirectObject`, a subclass of `PdfObject`, returns the referenced object here. Here we might get some errors if the PDF is malformed.
* We expect an `ArrayObject` here, which subclasses `list`. For this reason, I do not really see a reason for the `cast` (which is a no-op for regular execution anyway and only satisfies type checkers).

Reply:
It seems that I was being dragged down by the “cast” in the original code, and that the PDF 1.7 spec predicts an ArrayObject and allows looping without modification.
I am not sure of the advantage of IndirectObjects, but I will check it out when I fix other code.

Thanks again, this is very helpful!